internal/utils/metrics.go

```diff
@@ -0,0 +1,176 @@
+	commonMetrics
+	queueSize      *prometheus.GaugeVec
+	txCount        *prometheus.CounterVec
+	txPerBlock     prometheus.Histogram
+	blockSizeBytes prometheus.Histogram
+}
+
+func NewTxProcessingMetrics(reg *prometheus.Registry) *TxProcessingMetrics {
+	if reg == nil {
+		// If the registry is nil, then metrics collection is disabled
+		return nil
+	}
+	const namespace = "tx_processing"
+	s := &TxProcessingMetrics{
+		commonMetrics: *newCommonMetrics(namespace, reg),
+		queueSize: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "queue_size_txs",
+		}, []string{"queue_type"}),
+		txCount: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "tx_count",
+		}, []string{"validation_code", "transaction_type"}),
+		txPerBlock: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Namespace: namespace,
+			Name:      "tx_per_block",
+			Help:      "The number of transactions per block",
+			Buckets:   SizeBase10Buckets,
+		}),
+		blockSizeBytes: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Namespace: namespace,
+			Name:      "block_size_bytes",
+			Help:      "Total block size in bytes",
+			Buckets:   SizeBase2Buckets,
+		}),
+	}
+	reg.MustRegister(
+		s.queueSize,
+		s.txCount,
+		s.txPerBlock,
+		s.blockSizeBytes,
+	)
+	return s
+}
+
+func (s *TxProcessingMetrics) Latency(label string, startTime time.Time) {
+	// We redefine this method to prevent segmentation fault
+	if s != nil {
+		s.commonMetrics.Latency(label, startTime)
+	}
+}
+
+func (s *TxProcessingMetrics) QueueSize(label string, size int) {
+	if s != nil {
+		s.queueSize.WithLabelValues(label).Set(float64(size))
+	}
+}
+
+func (s *TxProcessingMetrics) IncrementTxCount(flag types.Flag, txType string) {
+	if s != nil {
+		s.txCount.WithLabelValues(flag.String(), txType).Inc()
+	}
+}
+
+func (s *TxProcessingMetrics) TxPerBlock(size int) {
+	if s != nil {
+		s.txPerBlock.Observe(float64(size))
+	}
+}
+
+func (s *TxProcessingMetrics) BlockSize(size int64) {
+	if s != nil {
+		s.blockSizeBytes.Observe(float64(size))
+	}
+}
+
+type DataRequestHandlingMetrics struct {
+	commonMetrics
+	txSize prometheus.Histogram
+}
+
+func NewDataRequestHandlingMetrics(reg *prometheus.Registry) *DataRequestHandlingMetrics {
+	if reg == nil {
+		// If the registry is nil, then metrics collection is disabled
+		return nil
+	}
+	const namespace = "data_handling"
+	s := &DataRequestHandlingMetrics{
+		commonMetrics: *newCommonMetrics(namespace, reg),
+		txSize: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Namespace: namespace,
+			Name:      "tx_size_bytes",
+			Help:      "Total TX size in bytes",
+			Buckets:   SizeBase2Buckets,
+		}),
+	}
+	reg.MustRegister(s.txSize)
+	return s
+}
+
+func (s *DataRequestHandlingMetrics) Latency(label string, startTime time.Time) {
+	// We redefine this method to prevent segmentation fault
+	if s != nil {
+		s.commonMetrics.Latency(label, startTime)
+	}
+}
```

Comment:
For some reason, when accessing a parent class method, it tries to dereference the object.
It seems weird to me because the correct method call can be resolved at compile time, but it is what it is.
So I resolved this by redefining the method in the child class.